SNT/TypeGenerator.tt

```diff
@@ -0,0 +1,257 @@
+<#@ template language="C#" #>
+<#@ assembly name="System.Core" #>
+<#@ import namespace="System.Linq" #>
+<#@ import namespace="System.Text" #>
+<#@ import namespace="System.Collections.Generic" #>
+using System;
+using System.Buffers;
+using System.Linq;
+using System.Numerics.Tensors;
+using System.Runtime.CompilerServices;
+using TorchSharp;
+
+namespace Torch.SNT 
+{
+<#
+foreach (var type in new [] { 
+    new { ApiName = "Byte", Storage = "byte", AccReal="long" },
+    new { ApiName = "Short", Storage = "short", AccReal="long" },   
+    new { ApiName = "Int", Storage = "int",  AccReal="long"},
+    new { ApiName = "Long", Storage = "long", AccReal="long" },
+    new { ApiName = "Double", Storage = "double", AccReal="double" },
+    new { ApiName = "Float", Storage = "float", AccReal="double" },
+     }) {
+        string tname = type.ApiName;
+        string stname = type.Storage;
+#>
+    /// <summary>
+    ///   Wrapper class used to surface a Torch <#=tname#>Tensor as a System.Numerics DensorTensor of <#=stname#>
+    /// </summary>
+    public sealed class <#=tname#>TorchTensor : DenseTensor<<#=stname#>>, IDisposable
+    {
+        internal sealed class <#=tname#>NativeMemory : MemoryManager<<#=stname#>>
+        {
+            private readonly <#=tname#>Tensor.<#=tname#>Storage storage;
+
+            public <#=tname#>NativeMemory(<#=tname#>Tensor.<#=tname#>Storage storage)
+            {
+                this.storage = storage;
+
+                if (storage.Size () < 0)
+                {
+                    throw new ArgumentOutOfRangeException ("Length cannot be negative.");
+                }
+            }
+
+            /// <summary>
+            /// Destructor for the native memory.
+            /// The Tensor memory lifecycle is managed by Torch. Disposing the memory
+            /// from here will through a "double free or corruption" error.
+            /// <summary>
+            ~<#=tname#>NativeMemory ()
+            {
+                Dispose (true);
```

Comment:
Shouldn't this be false?

Reply:
In this version I actually have to dispose the memory. If you look at line 146, I am retain the storage pointer every time a new native memory is created (or pinned). Therefore I need to `Dispose (true)` to properly decrement the reference count. I have to fix the comment :)

Reply:
I don't believe you can safely do that.  The GC does not guarantee that `storage` (another managed object) has not yet been finalized at the time the finalizer for this NativeMemory runs.  Only one object should be in charge of maintaining the native state and that object should be the one that has the finalizer that does work.  (ps, consider https://github.com/xamarin/TorchSharp/pull/51#discussion_r234733245)

Reply:
There is something I am missing on C# GC then. Native storage maintains a reference to `storage`, why the GC should destroy `storage` if refcount > 0?

Reply:
Both references are out of scope at the same time.  This is [documented](https://docs.microsoft.com/en-us/dotnet/api/system.object.finalize) behavior

>The finalizers of two objects are not guaranteed to run in any specific order, even if one object refers to the other. That is, if Object A has a reference to Object B and both have finalizers, Object B might have already been finalized when the finalizer of Object A starts.

I suspect this has to do with GC performance and how it calculates live objects in a single pass and puts them on the finalization queue in the order they are encountered (not necessarily in the order they appear in an object graph), also likely has to do with not needing to deal with potentially deep object cycles.  @Maoni0 could explain better.

Reply:
That sounds right.  Of course in the case that no-one calls Dispose, the finalizer for storage will still clean up the native resources, therefore you can technically remove the finalizer from NativeMemory since it does nothing.  You can also remove the public Dispose method unless you need to explicitly call it, since the base type implements IDisposable.

Did you consider unifying the storage and NativeMemory classes?  I think that actually might reduce the number of object allocations per tensor and feels like the right conceptual pairing.

Reply:
> Did you consider unifying the storage and NativeMemory classes? I think that actually might reduce the number of object allocations per tensor and feels like the right conceptual pairing.

Yes I was looking into that but this will require to add a dependency to the main TorchSharp project which I believe @migueldeicaza and @markusweimer want to avoid.

Reply:
>  this will require to add a dependency to the main TorchSharp project

You're talking about a dependency on System.Memory?  Seems like that's not such a bad depdendency to have since it would let TorchSharp use Span, but I can see why you may want to avoid it.

I think this is leading to confusion in how this NativeMemory class is implemented.  I was taking a closer look and now I see that Dispose is calling Free, which but I don't think this is correct.  It looks to me like you're using Free to represent "decrement a reference count" however the reference count may not have been previously incremented since I don't see a case where this is happening during construction.  Since the NativeMemory itself isn't the owner of the memory at all, I wonder if it even makes sense to do anything in its Dispose 😕: in fact I don't even see where you call this.

In the Tensor/NativeMemory sample we were using the NativeMemory object as the "owner" of the memory.  As such, it was the one implementing the reference count and it was the one responsible for eventually freeing that memory in its finalizer/dispose.

In this case, you're using NativeMemory as an adapter for your storage class which is ultimately the owner of the memory.  The adapter acts as the mechanism for constructing spans over the storage.

Furthermore you have added IDisposable to your derived Tensors and it looks to me like this will dispose the backing torch tensor which will dispose the backing storage.  I would have imagined you'd instead want to have decremented a reference count here and dispose the backing storage when the ref count gets to zero.  As its implemented I think you'll end up freeing the memory behind a shallow-copied reshaped tensor.

Perhaps you can add some test cases to further flesh out how this should behave?   

@tannergooding perhaps we could expand our Tensor samples to cover this scenario?

Reply:
> It looks to me like you're using Free to represent "decrement a reference count" however the reference count may not have been previously incremented since I don't see a case where this is happening during construction.

I have pushed the call to Retain into the Nativ Memory constructor. Before it was outside when the storage was retrieved from the tensor to create the NativeMemory object.

> Since the NativeMemory itself isn't the owner of the memory at all, I wonder if it even makes sense to do anything in its Dispose 😕: in fact I don't even see where you call this.

Dispose is required because NameMemory extends MemoryManager. 

> Furthermore you have added IDisposable to your derived Tensors and it looks to me like this will dispose the backing torch tensor which will dispose the backing storage. I would have imagined you'd instead want to have decremented a reference count here and dispose the backing storage when the ref count gets to zero. As its implemented I think you'll end up freeing the memory behind a shallow-copied reshaped tensor.

This is exactly the behavior I implemented (I hope :) ). The fact is that NativeMemory is required to dispose the storage because it may happen the case where a span is retrieved and used outside a tensor context (e.g., in [this](https://github.com/xamarin/TorchSharp/pull/51/files#diff-9d09654dea42c82f2b91e9afc1c739a1R1881) test case).